### PR TITLE
os: let ObjectStore::create() return unique_ptr<> 

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-26_11:04:40-rados-wip-kefu-testing-2021-05-26-1620-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-27_05:52:27-rados-wip-kefu-testing-2021-05-27-1155-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-27_09:22:11-rados-wip-kefu-testing-2021-05-27-1528-distro-basic-smithi/

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -2115,8 +2119,6 @@ int OSD::mkfs(CephContext *cct, ObjectStore *store, uuid_d fsid, int whoami, str
     ch.reset();
   }
   store->umount();
-free_store:
```
Comment (rzarzynski) [inline src/osd/OSD.cc:2118]: :+1:

File: src/tools/ceph_objectstore_tool.cc [hunk 2]
```diff
@@ -3601,7 +3601,16 @@ int main(int argc, char **argv)
     }
     return 1;
   }
-
+  auto umount_fs = make_scope_guard([&] {
```
Comment (rzarzynski) [inline src/tools/ceph_objectstore_tool.cc:3604]: I'm afraid this commit isn't just a refactoring. In the very end we have an extra logic converting the return value:

```cpp
  if (ret < 0)
    ret = 1;
  return ret;
}
```

Would it take effect after the change?

File: src/osd/OSD.h [hunk 3]
```diff
@@ -100,7 +100,7 @@ class OSDService {
   CephContext *cct;
   ObjectStore::CollectionHandle meta_ch;
   const int whoami;
-  ObjectStore *&store;
```
Comment (rzarzynski) [inline src/osd/OSD.h:103]: :+1:

File: src/tools/ceph_objectstore_tool.cc (hunk 2 above)
Comment (tchaikov) [reply src/tools/ceph_objectstore_tool.cc]: @rzarzynski  right. that part is difficult. i reverted this change. could you take another look?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-28_02:36:20-rados-wip-kefu-testing-2021-05-28-0853-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-28_13:33:45-rados-wip-kefu-testing-2021-05-28-1806-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-29_02:33:40-rados-wip-kefu-testing-2021-05-28-1806-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/45423
- https://tracker.ceph.com/issues/50151

**Pull request merged**